Upgrade Firefox flatpak to latest Firefox release

andrunko opened the pull request:
This is a PR to follow up on the changes to upgrade the Firefox flatpak to latest Firefox release. This PR is  based on the work done by @refi64 at https://github.com/refi64/firefox-flatpak/tree/updates.

https://phabricator.endlessm.com/T22543

andrunko commented:
First off, some general comments:
- While I understand some of these changes come from syncing with the upstream work being done at https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/, I think we should split the changes into actual build fixes/improvements and then other changes such as changes in strings/translations/metadata and we can then decide which ones to keep (e.g. our translations vs upstream translations).
- That said, I would suggest we split the changes (with each point below in its own commit) as follows (or close to):
  - build fixes/updates/improvements:
    - Port existing json as is to yaml without actual changes (and remove json)
    - Rename `endless-default-prefs.js` to `default-preferences.js` to be more aligned with upstream
    - Apply build system changes to build the new version and keep extra data sources inline in yaml
      - I would probably keep the changes to `apply_extra` inline as well (they are simple enough) and remove the new `apply_extra.sh`
    - Move extra data sources to separate files (e.g. `firefox-sources.json`) as done in the current changes
    - Add/update preferences (i.e. `endless-default-prefs.js` and `distribution.ini`)
    - Changes to `update-firefox-version` script
    - Add flash plugin extension
    - Remove unneeded files (e.g. `firefox-plugins-installer`)
    - changes to .gitignore
  - updates to strings/translations/metadata in general:
    - Sync with upstream but keep our extra metadata (e.g. content_rating for oars)
      - Possibly split the sync changes also into smaller chunks (e.g. update translations, update screenshots, add extra metadata, etc)
- Optional: Add `README.md` with some basic instructions to build (e.g. install runtime deps, build steps)

May be worth mentioning in the commit msgs where the changes come from, for example something like:
```
Sync translations with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

(synced with upstream commit <hash>)
```

Generally, all changes that we make that are valid for the upstream work we should also propose/discuss with upstream for integration after we are done here.

More specific comments to follow.

andrunko commented on apply_extra.sh:
```diff
@@ -0,0 +1,13 @@
+tar xf firefox-*.tar.bz2
+rm firefox-*.tar.bz2
+
+ln -s /app/firefox/plugins/plugins firefox/browser
```
Typo in `plugins/plugins`? Also `firefox/browser` doesn't seem to point to any plugins.

refi64 replied:
Not quite a typo, the plugin is set to:

```
  org.mozilla.Firefox.Plugin:
    add-ld-path: lib
    directory: firefox/plugins
    merge-dirs: plugins
    subdirectories: true
    no-autodownload: true
```

So, the file structure under `/app/firefox/plugins` will look like:

```
/app/firefox/plugins/
  Flash/    # an individual plugin
  plugins/  # the merge-dirs argument, has the contents of Flash/plugins and any others merged together
```

Firefox will read plugins by default from `browser-directory/browser/plugins`, so I'm essentially symlinking the merged plugin directory there. I can add a comment explaining this if you prefer.

andrunko replied:
I see, thanks - I think it would be good to add a comment but not a requirement.

andrunko commented on distribution.ini:
```diff
@@ -0,0 +1,8 @@
+
+[Preferences]
+app.update.enabled=false
+browser.shell.checkDefaultBrowser=false
```
Not sure we want to disable `checkDefaultBrowser` by default - our chrome has it enabled. @wjt? I understand this was also done upstream but maybe we should keep it enabled in our version.

refi64 replied:
I was under the impression that upstream did it because the default browser check also offers to *change* the default browser, which doesn't work inside a Flatpak afaik.

ramcq replied:
Yeah, that needs a default apps portal which I think atm does not exist.

andrunko replied:
All right, I see.

andrunko commented on endless-default-prefs.js:
```diff
@@ -3,3 +3,5 @@ pref("intl.locale.matchOS", true);
 pref("app.update.enabled", false);
 pref("app.update.silent", false);
+pref("app.update.autoInstallEnabled", false);
+pref("browser.shell.checkDefaultBrowser", false);
```
Same here regarding default browser check.

ramcq replied:
One thing unclear to me is whether we need both these files, if we have a distribution.ini?

andrunko replied:
Good point, I actually wondered the same but forgot to mention (also upstream has both files fwiw).

andrunko commented on org.mozilla.Firefox.appdata.xml:
```diff
@@ -49,9 +46,9 @@
-    </screenshot>
+    <screenshot type="default">https://raw.githubusercontent.com/hughsie/fedora-appstream/master/screenshots-extra/firefox/a.png</screenshot>
+    <screenshot>https://raw.githubusercontent.com/hughsie/fedora-appstream/master/screenshots-extra/firefox/b.png</screenshot>
+    <screenshot>https://raw.githubusercontent.com/hughsie/fedora-appstream/master/screenshots-extra/firefox/c.png</screenshot>
```
Wonder if we should keep the screenshot of it running on EndlessOS instead - and maybe update it to the new version.

Anyway, this is a detail we can check later, but please do it as a separate commit as mentioned before and we can then decide whether to include this change.

andrunko commented on org.mozilla.Firefox.appdata.xml:
```diff
@@ -65,7 +62,11 @@
+
+  <url type="homepage">https://www.mozilla.org/firefox/</url>
+  <url type="donation">https://donate.mozilla.org/</url>
+  <url type="bugtracker">https://bugzilla.mozilla.org/</url>
```
I think we should keep the previous bugtracker url here as it points specifically to the firefox bugtracker (and suggest the change upstream also).

andrunko commented on org.mozilla.Firefox.appdata.xml:
```diff
@@ -65,7 +62,11 @@
+  <url type="homepage">https://www.mozilla.org/firefox/</url>
+  <url type="donation">https://donate.mozilla.org/</url>
+  <url type="bugtracker">https://bugzilla.mozilla.org/</url>
+  <url type="help">https://support.mozilla.org/</url>
```
This entry is duplicated below, I would probably use this new URL here though given that the old URL is en-US specific.

andrunko commented on org.mozilla.Firefox.desktop:
```diff
@@ -1,34 +1,271 @@
 Icon=org.mozilla.Firefox
 Terminal=false
+Type=Application
+MimeType=text/html;text/xml;application/xhtml+xml;application/vnd.mozilla.xul+xml;text/mml;x-scheme-handler/http;x-scheme-handler/https;-X-Endless-LaunchMaximized=true;
```
`X-Endless-LaunchMaximized` should go to its own line and without a `;` at the end.

andrunko commented on plugins/org.mozilla.Firefox.Plugin.Flash.yaml:
```diff
@@ -0,0 +1,39 @@
+      - type: extra-data
+        filename: flash_player_npapi_linux.tar.gz
+        only-arches: [x86_64]
+        url: https://fpdownload.adobe.com/pub/flashplayer/pdc/32.0.0.207/flash_player_npapi_linux.x86_64.tar.gz
```
This points to a broken URL - see https://get.adobe.com/flashplayer/webservices/json/?platform_type=Linux&platform_arch=x86-64&eventname=flashplayerotherversions&browser_dist=Firefox for latest.

andrunko commented on plugins/org.mozilla.Firefox.Plugin.Flash.yaml:
```diff
@@ -0,0 +1,39 @@
+      - type: extra-data
+        filename: flash_player_npapi_linux.tar.gz
+        only-arches: [i386]
+        url: https://fpdownload.adobe.com/pub/flashplayer/pdc/32.0.0.207/flash_player_npapi_linux.i386.tar.gz
```
Same as above.

andrunko commented on update-firefox-version:
```diff
@@ -22,36 +22,75 @@
-class FirefoxJsonGenerator():
-    JSON_FILE_NAME = 'org.mozilla.Firefox.json.in'
-    MOZILLA_ARCHIVE_URL = 'https://archive.mozilla.org/pub/firefox/releases/{version}/linux-x86_64/'
+CURRENT_VERSION = '60.7.1esr'
```
I checked with @ramcq / @wjt and we agreed to go with non-esr version here, so please update. Current latest is `68.0.1`.

andrunko commented on plugins/org.mozilla.Firefox.Plugin.Flash.yaml:
```diff
@@ -0,0 +1,39 @@
+app-id: org.mozilla.Firefox.Plugin.Flash
```
Wonder if we should use the same name used by the unofficial firefox flatpak at https://github.com/ozoder/firefox-devedition-flatpak/tree/master/com.adobe.FlashPlayer.NPAPI.

Also add missing appdata for it (as done in the repo above).

refi64 replied:
Worth noting that the common prefix has to be shared by all plugins, so if e.g. there were another plugin that were to be provided via this path (e.g. codecs or something), they would also need to have the Flash-specific prefix.

ramcq replied:
That might be over-generalisation considering now I think the only NPAPI plugin that Firefox would ever load is Flash, and only for maybe 1-2 years more. So actually hardcoding the Firefox flatpak to a specific Flash extension point might be fine.

andrunko commented on firefox.flatpakref.in:
```diff
@@ -7,5 +7,5 @@ IsRuntime=False
 
 Comment=firefox
-Description=Firefox web browser
+Description=Firefox Web Browser
```
The link for Icon below is broken.

andrunko commented on org.mozilla.Firefox.appdata.xml:
```diff
@@ -1,13 +1,10 @@
-  <categories>
-    <category>Utility</category>
-  </categories>
-
```
Why delete this?

refi64 replied:
Apparently now the preferred way is to leave categories inside the .desktop file, and they'll automatically be pulled from there instead of duplicating them here too.

ramcq replied:
Yeah, appstream-compose handles this.

andrunko replied:
Ack.

andrunko commented:
Also another important thing. Given the flash extension depends on a new flatpak version (with support for extra-data in extensions) which is currently not available on EndlessOS, please do the changes to use the extension separately such that we can still use the flash installer while we dont update the flatpak version to be able to use the extension.

andrunko commented:
> Also another important thing. Given the flash extension depends on a new flatpak version (with support for extra-data in extensions) which is currently not available on EndlessOS, please do the changes to use the extension separately such that we can still use the flash installer while we dont update the flatpak version to be able to use the extension.

IOTW, please keep the flash plugin installer functionality working with the new firefox version up until the commit where you move to use the extension.

ramcq commented:
> > Also another important thing. Given the flash extension depends on a new flatpak version (with support for extra-data in extensions) which is currently not available on EndlessOS, please do the changes to use the extension separately such that we can still use the flash installer while we dont update the flatpak version to be able to use the extension.
> 
> IOTW, please keep the flash plugin installer functionality working with the new firefox version up until the commit where you move to use the extension.

The problem is not that we could rely on the new functionality, the problem is sequencing the upgrades. Having a single cut-over point in our git history doesn't solve ordering the OS and the Flatpak updates which arrive independently - until we've released the OS update with the new Flatpak, any current Endless OS user (with old Flatpak) could fetch this Firefox Flatpak, and it would probably break Flash for them, if it didn't fail / spit some errors.

So - is there any way in a Flatpak to flag a minimum version number that will otherwise block the update? Otherwise, we could simply keep the installer script as a fallback - add an early exit to do nothing (or even remove the ~ plugin) in case that the extension one is present. That would work in either case (new OS first, then new Flatpak, versus new Flatpak first, then new OS later).

ramcq commented on distribution.ini:
```diff
@@ -0,0 +1,8 @@
+[Global]
+id=mozilla-flatpak
```
Do we need to change this given it's not in fact Mozilla's Flatpak if it comes from Endless? :)

ramcq commented on distribution.ini:
```diff
@@ -0,0 +1,8 @@
+[Global]
+id=mozilla-flatpak
+version=1.0
+about=Mozilla Firefox Flatpak
```
Similar. Lies? :)

refi64 commented:
AFAIK the only way to block an upgrade would be `--require-version`, which I'm already doing for the actual Flash plugin, but that would prevent people with the older version from upgrading Firefox *at all*, not just to what came before the latest version. (Though I guess if the person hasn't updated their host system they probably wouldn't be updating the Flatpaks themselves either?)

refi64 committed:
Avoid modifying the primary json file

This generates the extra-data sources to a separate file instead of overwriting the original.

It also drops the BeautifulSoup4 depedency of the updater script.

refi64 committed:
Convert JSON manifest to YAML

refi64 committed:
Migrate to fd.o version 18.08

refi64 committed:
Update Firefox to non-ESR 68.0.1

refi64 committed:
Revamp Makefile to be more scalable

refi64 committed:
Revamp manifest in sync with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Drop the Java plugin

refi64 committed:
Use an extra-data-based extension for Flash player

refi64 committed:
Update prefs and add distribution metadata in sync with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Update gitignore

refi64 committed:
Modernize appdata file

refi64 committed:
Update flatpakref icon url

refi64 committed:
Update capitalization based on upstream data

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Update appdata URLs in sync with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Update appdata screenshots in sync with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Update desktop metadata in sync with upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Add desktop actions based on upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 committed:
Update desktop file translations based on upstream

Upstream URL: https://hg.mozilla.org/projects/maple/file/tip/taskcluster/docker/firefox-flatpak/

Commit: 278f506dd3abcabc076e8b9da1ec6756b8a6b8bd

refi64 commented:
Okay I split everything up as requested and AFAIK addressed all the review feedback.

refi64 committed:
Add Flash appdata

refi64 committed:
Convert YAML to JSON for easier interop

Commit:
Remove YAML manifests

Let's not keep both formats otherwise we have to keep them
in sync (extra maintenance work).
Favoring JSON for easier interop with existing CI integration
scripts.

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

andrunko commented:
Thanks @refi64, all good, merging!
